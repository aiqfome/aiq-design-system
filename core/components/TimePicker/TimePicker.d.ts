import React from 'react';
export interface Props {
    value?: any;
    sufix?: any;
    name?: string;
    label?: string;
    errorForm?: boolean;
    hasSeconds?: boolean;
    placeholder?: string;
    errorMessage?: string;
    type?: 'minSec' | 'all';
    maxWidth?: string | number;
    onChange?: (e: any) => void;
    variant?: 'outlined' | 'default';
    getValue?: (input: any) => string;
}
export declare const TimePicker: React.FC<Props>;
//# sourceMappingURL=TimePicker.d.ts.map