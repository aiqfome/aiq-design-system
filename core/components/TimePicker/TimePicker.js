"use strict";
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.TimePicker = void 0;
const react_1 = __importDefault(require("react"));
const prop_types_1 = __importDefault(require("prop-types"));
const TimePickerAll_1 = require("./TimePickerAll");
const TimePickerMinSec_1 = require("./TimePickerMinSec");
exports.TimePicker = (_a) => {
    var { type = 'all' } = _a, props = __rest(_a, ["type"]);
    if (type === 'minSec') {
        return react_1.default.createElement(TimePickerMinSec_1.TimePickerMinSec, Object.assign({}, props));
    }
    return react_1.default.createElement(TimePickerAll_1.TimePickerAll, Object.assign({}, props));
};
exports.TimePicker.displayName = 'TimePicker';
exports.TimePicker.propTypes = {
    sufix: prop_types_1.default.any,
    name: prop_types_1.default.string,
    label: prop_types_1.default.string,
    onChange: prop_types_1.default.func,
    getValue: prop_types_1.default.func,
    errorForm: prop_types_1.default.bool,
    hasSeconds: prop_types_1.default.bool,
    placeholder: prop_types_1.default.string,
    errorMessage: prop_types_1.default.string,
    type: prop_types_1.default.oneOf(['minSec', 'all']),
    variant: prop_types_1.default.oneOf(['outlined', 'default']),
    maxWidth: prop_types_1.default.oneOfType([prop_types_1.default.string, prop_types_1.default.number])
};
