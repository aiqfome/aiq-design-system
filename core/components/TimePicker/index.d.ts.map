{"version":3,"file":"index.d.ts","sourceRoot":"","sources":["../../../src/components/TimePicker/index.tsx"],"names":[],"mappings":"AAAA,cAAc,cAAc,CAAA"}