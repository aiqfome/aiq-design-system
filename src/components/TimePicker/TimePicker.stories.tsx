import React, { ReactElement } from 'react'

import { TimePicker } from './TimePicker'

import { Flex } from '../Flex'

export default {
  component: TimePicker,
  title: 'TimePicker'
}

export const Basic: React.FC = (): ReactElement => (
  <Flex variant='fullCentralized'>
    <TimePicker
      onChange={value => console.log(value)}
      placeholder='aiqfome'
      maxWidth='100px'
      value='12:30'
    />
  </Flex>
)

export const Outlined: React.FC = (): ReactElement => (
  <Flex variant='fullCentralized'>
    <TimePicker variant='outlined' label='aiqfome' maxWidth='120px' />
  </Flex>
)
