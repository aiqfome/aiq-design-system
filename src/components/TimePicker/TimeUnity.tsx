import React, { useState, useEffect } from 'react'

import PropTypes from 'prop-types'

import styled from 'styled-components'

import { MdExpandLess, MdExpandMore } from 'react-icons/md'

import { Flex } from '../Flex'
import { Icon } from '../Icon'
import { Text } from '../Text'

interface Props {
  max: number
  min: number
  label: string
  value: number
  onChange?: (value: any) => void
}

interface ButtonProps {
  onClick: () => void
}

const Button = styled(Icon)<ButtonProps>`
  &:hover {
    cursor: pointer;
  }
`

export const TimeUnity: React.FC<Props> = ({
  max,
  min,
  label,
  value,
  onChange = () => {
    // do nothing.
  }
}) => {
  const [number, setNumber] = useState(0)

  useEffect(() => {
    setNumber(value)
  }, [value])

  function upNumber() {
    if (max > number) {
      setNumber(number + 1)
      onChange(('00' + (number + 1)).slice(-2))
    } else {
      setNumber(min)
    }
  }

  function downNumber() {
    if (min < number) {
      setNumber(number - 1)
      onChange(('00' + (number - 1)).slice(-2))
    } else {
      setNumber(max)
    }
  }

  return (
    <Flex variant='centralized' flexDirection='column'>
      <Button onClick={upNumber} color='primary'>
        <MdExpandLess size={32} />
      </Button>
      <Text fontSize='xlarge' fontWeight='medium'>
        {('00' + number).slice(-2)}
      </Text>
      <Button onClick={downNumber} color='primary'>
        <MdExpandMore size={32} />
      </Button>
      <Text color='primary' fontSize='small' fontWeight='medium'>
        {label}
      </Text>
    </Flex>
  )
}

TimeUnity.propTypes = {
  max: PropTypes.number.isRequired,
  min: PropTypes.number.isRequired,
  label: PropTypes.string.isRequired,
  value: PropTypes.number.isRequired,
  onChange: PropTypes.func
}
