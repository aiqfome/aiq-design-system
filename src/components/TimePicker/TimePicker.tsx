import React, { useState, useEffect } from 'react'
import PropTypes from 'prop-types'

import styled from 'styled-components'

import { Input } from '../Input'
import { Box } from '../Box'
import { Flex } from '../Flex'
import { TimeUnity } from './TimeUnity'

export interface Props {
  value?: any
  placeholder?: string
  label?: string
  variant?: 'outlined' | 'default'
  onChange?: (e: any) => void
  maxWidth?: string | number
}

interface PickerProps {
  onMouseEnter: () => void
  onMouseLeave: () => void
}

const Picker = styled(Box)<PickerProps>``

export const TimePicker: React.FC<Props> = ({
  value,
  placeholder,
  label,
  variant,
  maxWidth,
  onChange = (e: any) => {
    // do nothing
  },
  ...props
}) => {
  const [showPicker, setShowPicker] = useState(false)
  const [inputValue, setInputValue] = useState(value || '')

  useEffect(() => {
    onChange(inputValue)
  }, [inputValue])

  function applyMask(value) {
    return value
      .replace(/[\D]+/g, '')
      .replace(/(\d{2})(\d)/, '$1:$2')
      .replace(/(:\d{2})\d+?$/, '$1')
  }

  function handleInputOnChange(e) {
    setInputValue(applyMask(e.target.value))
  }

  function getHour() {
    const valueSplited = inputValue.split(':')
    if (valueSplited.length > 0) {
      return parseInt(valueSplited[0] || 0)
    }
    return 0
  }

  function getMin() {
    const valueSplited = inputValue.split(':')
    if (valueSplited.length > 1) {
      return parseInt(valueSplited[1] || 0)
    }
    return 0
  }

  function handleOnChangeHour(hour) {
    const valueSplited = inputValue.split(':')
    if (valueSplited.length > 0) {
      valueSplited[0] = hour
      if (valueSplited.length < 2) {
        valueSplited[1] = '00'
      }
      setInputValue(`${valueSplited[0]}:${valueSplited[1]}`)
    }
  }

  function handleOnChangeMinute(minute) {
    const valueSplited = inputValue.split(':')
    if (valueSplited.length > 1) {
      valueSplited[1] = minute
      setInputValue(`${valueSplited[0]}:${valueSplited[1]}`)
    }
  }

  return (
    <Picker
      position='relative'
      maxWidth={maxWidth}
      onMouseEnter={() => setShowPicker(true)}
      onMouseLeave={() => setShowPicker(false)}
      {...props}
    >
      <Input
        value={inputValue}
        variant={variant}
        placeholder={placeholder}
        label={label}
        onChange={handleInputOnChange}
      />
      {showPicker && (
        <Box
          width='100%'
          maxWidth={maxWidth}
          position='absolute'
          top='38px'
          border='1px solid #dedede'
          boxShadow='0px 3px 6px #00000017'
          borderRadius='8px'
          padding='8px 16px'
        >
          <Flex flexDirection='row' justifyContent='space-between'>
            <TimeUnity
              value={getHour()}
              onChange={handleOnChangeHour}
              label='hora'
              max={23}
              min={0}
            />
            <TimeUnity
              value={getMin()}
              onChange={handleOnChangeMinute}
              label='min'
              max={59}
              min={0}
            />
          </Flex>
        </Box>
      )}
    </Picker>
  )
}

TimePicker.propTypes = {
  value: PropTypes.any,
  variant: PropTypes.oneOf(['outlined', 'default']),
  onChange: PropTypes.func,
  placeholder: PropTypes.string,
  label: PropTypes.string,
  maxWidth: PropTypes.oneOfType([PropTypes.string, PropTypes.number])
}
