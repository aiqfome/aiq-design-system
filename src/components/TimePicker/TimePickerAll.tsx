import React, { useState } from 'react'
import PropTypes from 'prop-types'

import styled from 'styled-components'

import { Box } from '../Box'
import { Flex } from '../Flex'
import { Input } from '../Input'
import { TimeUnity } from './TimeUnity'

export interface Props {
  value?: any
  sufix?: any
  name?: string
  label?: string
  errorForm?: boolean
  hasSeconds?: boolean
  placeholder?: string
  errorMessage?: string
  maxWidth?: string | number
  onChange?: (e: any) => void
  variant?: 'outlined' | 'default'
}

interface PickerProps {
  onMouseEnter: () => void
  onMouseLeave: () => void
}

const Picker = styled(Box)<PickerProps>``

export const TimePickerAll = React.forwardRef(
  <HTMLInputElement, Props>(
    {
      name,
      value,
      label,
      sufix,
      variant,
      maxWidth,
      errorForm,
      hasSeconds,
      placeholder,
      errorMessage,
      onChange = (e: any) => {
        console.log('input:', e)
      },
      ...props
    },
    ref
  ) => {
    const [showPicker, setShowPicker] = useState(false)
    const [inputValue, setInputValue] = useState(value || '')

    const applyMask = (value = '') => {
      if (hasSeconds) {
        return value
          .replace(/[\D]+/g, '')
          .replace(/(\d{2})(\d)/, '$1:$2')
          .replace(/(\d{2})(\d)/, '$1:$2')
          .replace(/(:\d{2})\d+?$/, '$1')
      }

      return value
        .replace(/[\D]+/g, '')
        .replace(/(\d{2})(\d)/, '$1:$2')
        .replace(/(:\d{2})\d+?$/, '$1')
    }

    const handleInputOnChange = e => {
      const { value = '' } = e?.target || {}
      const valueSplited = applyMask(value).split(':')

      if (valueSplited.length > 0 && valueSplited[0]) {
        const number = Number(valueSplited[0])
        if (number > 24) valueSplited[0] = '24'
      }

      if (valueSplited.length > 1 && valueSplited[1]) {
        const number = Number(valueSplited[1])
        if (number > 59) valueSplited[1] = '59'
      }

      if (valueSplited.length > 2 && valueSplited[2]) {
        const number = Number(valueSplited[2])
        if (number > 59) valueSplited[2] = '59'
      }

      const time = valueSplited.join(':')

      setInputValue(time)
      onChange(time)
    }

    const getTime = (time = '') => {
      const valueSplited = inputValue.split(':')

      if (time === 'hour' && valueSplited.length > 0) {
        return parseInt(valueSplited[0] || 0)
      }

      if (time === 'min' && valueSplited.length > 1) {
        return parseInt(valueSplited[1] || 0)
      }

      if (time === 'sec' && valueSplited.length > 2) {
        return parseInt(valueSplited[2] || 0)
      }

      return 0
    }

    const handleOnChangeTime = (time = '', value = '') => {
      const valueSplited = inputValue.split(':')
      const newValues = ['00', '00', '00']

      if (valueSplited.length > 0 && valueSplited[0]) {
        newValues[0] = valueSplited[0]
      }

      if (valueSplited.length > 1 && valueSplited[1]) {
        newValues[1] = valueSplited[1]
      }

      if (valueSplited.length > 2 && valueSplited[2]) {
        newValues[2] = valueSplited[2]
      }

      if (time === 'hour') newValues[0] = value
      if (time === 'min') newValues[1] = value
      if (time === 'sec') newValues[2] = value

      let formattedValue = `${newValues[0]}:${newValues[1]}`

      if (hasSeconds) {
        formattedValue = `${newValues[0]}:${newValues[1]}:${newValues[2]}`
      }

      setInputValue(formattedValue)
      onChange(formattedValue)
    }

    return (
      <Picker
        position='relative'
        maxWidth={maxWidth}
        data-testid='timepicker'
        onMouseEnter={() => setShowPicker(true)}
        onMouseLeave={() => setShowPicker(false)}
        {...props}
      >
        <Input
          name={name}
          label={label}
          sufix={sufix}
          inputRef={ref}
          variant={variant}
          value={inputValue}
          errorForm={errorForm}
          placeholder={placeholder}
          errorMessage={errorMessage}
          onChange={handleInputOnChange}
        />

        {showPicker && (
          <Box
            top='38px'
            zIndex={1}
            width='100%'
            padding='8px 16px'
            borderRadius='8px'
            position='absolute'
            maxWidth={maxWidth}
            backgroundColor='#fff'
            border='1px solid #dedede'
            data-testid='timepicker-open'
            boxShadow='0px 3px 6px #00000017'
          >
            <Flex flexDirection='row' justifyContent='space-between'>
              <TimeUnity
                min={0}
                max={23}
                label='hora'
                value={getTime('hour')}
                onChange={value => handleOnChangeTime('hour', value)}
              />

              <TimeUnity
                min={0}
                max={59}
                label='min'
                value={getTime('min')}
                onChange={value => handleOnChangeTime('min', value)}
              />

              {hasSeconds && (
                <TimeUnity
                  min={0}
                  max={59}
                  label='segundos'
                  value={getTime('sec')}
                  onChange={value => handleOnChangeTime('sec', value)}
                />
              )}
            </Flex>
          </Box>
        )}
      </Picker>
    )
  }
)

TimePickerAll.displayName = 'TimePickerAll'

TimePickerAll.propTypes = {
  value: PropTypes.any,
  sufix: PropTypes.any,
  name: PropTypes.string,
  label: PropTypes.string,
  onChange: PropTypes.func,
  errorForm: PropTypes.bool,
  hasSeconds: PropTypes.bool,
  placeholder: PropTypes.string,
  errorMessage: PropTypes.string,
  variant: PropTypes.oneOf(['outlined', 'default']),
  maxWidth: PropTypes.oneOfType([PropTypes.string, PropTypes.number])
}
