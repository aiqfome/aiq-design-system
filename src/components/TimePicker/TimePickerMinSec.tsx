import React, { useCallback, useState } from 'react'
import PropTypes from 'prop-types'

import styled from 'styled-components'

import { Box } from '../Box'
import { Flex } from '../Flex'
import { Input } from '../Input'
import { TimeUnity } from './TimeUnity'
import { useEffect } from 'react'

export interface Props {
  value?: any
  sufix?: any
  name?: string
  label?: string
  errorForm?: boolean
  placeholder?: string
  errorMessage?: string
  maxWidth?: string | number
  onChange?: (e: any) => void
  variant?: 'outlined' | 'default'
  onChangeInput?: (e: any) => void
  getValue?: (input: any) => string
}

interface PickerProps {
  onMouseEnter: () => void
  onMouseLeave: () => void
}

const Picker = styled(Box)<PickerProps>``

export const TimePickerMinSec = React.forwardRef(
  <HTMLInputElement, Props>(
    {
      name,
      value,
      label,
      sufix,
      variant,
      maxWidth,
      getValue,
      errorForm,
      placeholder,
      errorMessage,
      onChangeInput = (e: any) => {
        console.log('input:', e)
      },
      onChange = (e: any) => {
        console.log('input:', e)
      },
      ...props
    },
    ref
  ) => {
    const [showPicker, setShowPicker] = useState(false)
    const [inputValue, setInputValue] = useState(value || '')

    useEffect(() => {
      setInputValue(value || '')
    }, [value])

    const applyMask = (value = '') => {
      return value
        .replace(/[\D]+/g, '')
        .replace(/(\d{2})(\d)/, '$1:$2')
        .replace(/(:\d{2})\d+?$/, '$1')
    }

    const handleInputOnChange = e => {
      const { value = '' } = e?.target || {}
      if (getValue && onChangeInput) {
        onChangeInput(value)
        return ''
      }

      const valueSplited = applyMask(value).split(':')

      if (valueSplited.length > 0 && valueSplited[0]) {
        const number = Number(valueSplited[0])
        if (number > 59) valueSplited[0] = '59'
      }

      if (valueSplited.length > 1 && valueSplited[1]) {
        const number = Number(valueSplited[1])
        if (number > 59) valueSplited[1] = '59'
      }

      const time = valueSplited.join(':')

      setInputValue(time)
      onChange(time)
    }

    const getTime = (time = '') => {
      const valueSplited = inputValue.split(':')

      if (time === 'min' && valueSplited.length > 0) {
        return parseInt(valueSplited[0] || 0)
      }

      if (time === 'sec' && valueSplited.length > 1) {
        return parseInt(valueSplited[1] || 0)
      }

      return 0
    }

    const handleOnChangeTime = (time = '', value = '') => {
      const valueSplited = inputValue.split(':')
      const newValues = ['00', '00']

      if (valueSplited.length > 0 && valueSplited[0]) {
        newValues[0] = valueSplited[0]
      }

      if (valueSplited.length > 1 && valueSplited[1]) {
        newValues[1] = valueSplited[1]
      }

      if (time === 'min') newValues[0] = value
      if (time === 'sec') newValues[1] = value

      const formattedValue = `${newValues[0]}:${newValues[1]}`

      setInputValue(formattedValue)
      onChange(formattedValue)
    }

    const getInputValue = useCallback(() => {
      if (getValue) return getValue(inputValue)

      return inputValue
    }, [inputValue, getValue])

    return (
      <Picker
        maxWidth={maxWidth}
        position='relative'
        data-testid='timepicker'
        onMouseEnter={() => setShowPicker(true)}
        onMouseLeave={() => setShowPicker(false)}
        {...props}
      >
        <Input
          name={name}
          label={label}
          sufix={sufix}
          inputRef={ref}
          variant={variant}
          errorForm={errorForm}
          value={getInputValue()}
          placeholder={placeholder}
          errorMessage={errorMessage}
          onChange={handleInputOnChange}
        />

        {showPicker && (
          <Box
            top='38px'
            zIndex={1}
            width='100%'
            padding='8px 16px'
            borderRadius='8px'
            position='absolute'
            maxWidth={maxWidth}
            backgroundColor='#fff'
            border='1px solid #dedede'
            data-testid='timepicker-open'
            boxShadow='0px 3px 6px #00000017'
          >
            <Flex flexDirection='row' justifyContent='space-between'>
              <TimeUnity
                min={0}
                max={59}
                label='min'
                value={getTime('min')}
                onChange={value => handleOnChangeTime('min', value)}
              />

              <TimeUnity
                min={0}
                max={59}
                label='seg'
                value={getTime('sec')}
                onChange={value => handleOnChangeTime('sec', value)}
              />
            </Flex>
          </Box>
        )}
      </Picker>
    )
  }
)

TimePickerMinSec.displayName = 'TimePickerMinSec'

TimePickerMinSec.propTypes = {
  value: PropTypes.any,
  sufix: PropTypes.any,
  name: PropTypes.string,
  label: PropTypes.string,
  getValue: PropTypes.func,
  onChange: PropTypes.func,
  errorForm: PropTypes.bool,
  placeholder: PropTypes.string,
  onChangeInput: PropTypes.func,
  errorMessage: PropTypes.string,
  variant: PropTypes.oneOf(['outlined', 'default']),
  maxWidth: PropTypes.oneOfType([PropTypes.string, PropTypes.number])
}
