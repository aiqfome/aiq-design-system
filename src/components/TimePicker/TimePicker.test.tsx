import React from 'react'

import { render } from '../utils/test/render'

import { TimePicker } from './TimePicker'

describe('must match with the previous snapshot', () => {
  test('snapshot renders', () => {
    const component = render(
      <TimePicker
        onChange={value => console.log(value)}
        placeholder='aiqfome'
        maxWidth='120px'
        value='12:30'
      />
    )
    const tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })
})
