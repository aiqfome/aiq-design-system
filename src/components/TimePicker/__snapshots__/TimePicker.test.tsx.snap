// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`must match with the previous snapshot snapshot renders 1`] = `
<div
  className="sc-eCApGN gIxNaa sc-bqGHjH"
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
>
  <div
    className="sc-bdnylx jDchNm"
  >
    <input
      className="sc-iCoHVE kTQpRu"
      onChange={[Function]}
      placeholder="aiqfome"
      type="text"
      value="12:30"
    />
  </div>
</div>
`;
